hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -1674,6 +1693,111 @@ public void progressChanged(ProgressEvent progressEvent) {
     }
   }
 
+  protected void setSSEKMSOrCIfRequired(InitiateMultipartUploadRequest req) {
+    if (StringUtils.isNotBlank(serverSideEncryptionAlgorithm)){
+      if(S3AEncryptionMethods.SSE_KMS.getMethod()
+          .equals(serverSideEncryptionAlgorithm)) {
+        if (StringUtils.isNotBlank(serverSideEncryptionKey)) {
+          //Use specified key
+          req.setSSEAwsKeyManagementParams(
+              new SSEAwsKeyManagementParams(serverSideEncryptionKey)
+          );
+        }else{
+          //Use default key
+          req.setSSEAwsKeyManagementParams(new SSEAwsKeyManagementParams());
+        }
+      }else if(S3AEncryptionMethods.SSE_C.getMethod()
+          .equals(serverSideEncryptionAlgorithm)) {
+        if (StringUtils.isNotBlank(serverSideEncryptionKey)) {
+          //at the moment, only supports copy using the same key
+          req.setSSECustomerKey(new SSECustomerKey(serverSideEncryptionKey));
+        }
+      }
+    }
+  }
+
+
+  protected void setSSEKMSOrCIfRequired(CopyObjectRequest copyObjectRequest) {
+    if (StringUtils.isNotBlank(serverSideEncryptionAlgorithm)){
+      if(S3AEncryptionMethods.SSE_KMS.getMethod()
+          .equals(serverSideEncryptionAlgorithm)) {
+        if (StringUtils.isNotBlank(serverSideEncryptionKey)) {
+          //Use specified key
+          copyObjectRequest.setSSEAwsKeyManagementParams(
+              new SSEAwsKeyManagementParams(serverSideEncryptionKey)
+          );
+        }else{
+          //Use default key
+          copyObjectRequest.setSSEAwsKeyManagementParams(
+              new SSEAwsKeyManagementParams()
+          );
+        }
+      }else if(S3AEncryptionMethods.SSE_C.getMethod()
+          .equals(serverSideEncryptionAlgorithm)) {
+        if (StringUtils.isNotBlank(serverSideEncryptionKey)) {
+          //at the moment, only supports copy using the same key
+          copyObjectRequest.setSourceSSECustomerKey(
+              new SSECustomerKey(serverSideEncryptionKey)
+          );
+          copyObjectRequest.setDestinationSSECustomerKey(
+              new SSECustomerKey(serverSideEncryptionKey)
+          );
+        }
+      }
+    }
+  }
+
+  protected void setSSECIfRequired(GetObjectMetadataRequest request) {
+    if (StringUtils.isNotBlank(serverSideEncryptionAlgorithm)){
+      if(S3AEncryptionMethods.SSE_C.getMethod()
+          .equals(serverSideEncryptionAlgorithm)) {
+        if (StringUtils.isNotBlank(serverSideEncryptionKey)) {
+          //at the moment, only supports copy using the same key
+          request.setSSECustomerKey(
+              new SSECustomerKey(serverSideEncryptionKey)
```

Comment:
true, but not all of them can be merged. I'm relying in else clauses as well depending on some of the conditions being false. I'll try to rewrite it though and will see how it looks.